File: src/gtk4/items.c
// A control that offers a list of choices.

#include "internal.h"

static GtkStringList *ctd_item_list(gpointer object) {
    if (!GTK_IS_DROP_DOWN(object)) return NULL;
    GListModel *model = gtk_drop_down_get_model(GTK_DROP_DOWN(object));
    return model && GTK_IS_STRING_LIST(model) ? GTK_STRING_LIST(model) : NULL;
}

static ctd_status ctd_items_clear_raising(ctd_handle widget) {
    gpointer object = ctd_resolve(widget);
    if (!object) return CTD_ERR_STALE;
    GtkStringList *items = ctd_item_list(object);
    if (!items) return CTD_ERR_KIND;
    guint count = g_list_model_get_n_items(G_LIST_MODEL(items));
    if (count > 0) gtk_string_list_splice(items, 0, count, NULL);
    return CTD_OK;
}

// Silent, like every other write cortado makes: changing the model of a
// GtkDropDown moves its selection, and a moved selection notifies.
ctd_status ctd_items_clear(ctd_handle widget) {
    g_writing++;
    ctd_status status = ctd_items_clear_raising(widget);
    g_writing--;
    return status;
}

static ctd_status ctd_items_add_raising(ctd_handle widget, const char *utf8, int32_t len) {
    if (ctd_has_nul(utf8, len)) return CTD_ERR_RANGE;
    gpointer object = ctd_resolve(widget);
    if (!object) return CTD_ERR_STALE;
    GtkStringList *items = ctd_item_list(object);
    if (!items) return CTD_ERR_KIND;
    if (len < 0) return CTD_ERR_RANGE;
    char *text = ctd_dup(utf8, len);
    gtk_string_list_append(items, text);
    g_free(text);
    return CTD_OK;
}

// Silent, like every other write cortado makes: changing the model of a
// GtkDropDown moves its selection, and a moved selection notifies.
ctd_status ctd_items_add(ctd_handle widget, const char *utf8, int32_t len) {
    g_writing++;
    ctd_status status = ctd_items_add_raising(widget, utf8, len);
    g_writing--;
    return status;
}

ctd_status ctd_items_count(ctd_handle widget, int32_t *out) {
    gpointer object = ctd_resolve(widget);
    if (!object) return CTD_ERR_STALE;
    GtkStringList *items = ctd_item_list(object);
    if (!items) return CTD_ERR_KIND;
    if (out) *out = (int32_t)g_list_model_get_n_items(G_LIST_MODEL(items));
    return CTD_OK;
}

int32_t ctd_items_at(ctd_handle widget, int32_t index, char *out, int32_t cap) {
    gpointer object = ctd_resolve(widget);
    if (!object) return CTD_ERR_STALE;
    GtkStringList *items = ctd_item_list(object);
    if (!items) return CTD_ERR_KIND;
    guint count = g_list_model_get_n_items(G_LIST_MODEL(items));
    if (index < 0 || (guint)index >= count) return CTD_ERR_RANGE;
    return ctd_copy_out(gtk_string_list_get_string(items, (guint)index), out, cap);
}

File: src/mac/items.m
// A control that offers a list of choices.
//
// Items are replaced wholesale rather than patched — the reasoning is beside
// `ctd_items_clear` in ../cortado_host.h.

#import "internal.h"

// ------------------------------------------------------------------ item lists

// The control that holds the items. A combo box is its own list; a wrapper —
// a text area's scroll view, say — is not, and answers nil so the caller gets
// CTD_ERR_KIND rather than a silent no-op.
static NSPopUpButton *ctd_item_list(id object) {
    if ([object isKindOfClass:[NSPopUpButton class]]) return (NSPopUpButton *)object;
    return nil;
}

// A segmented control holds the same list and keeps it in a different place:
// AppKit has no menu behind it, only a segment count and a label per segment.
// Both answer ctd_items_*, because "a list of choices" is one idea and a
// caller should not have to know which control it landed in.
static NSSegmentedControl *ctd_item_bar(id object) {
    if ([object isKindOfClass:[NSSegmentedControl class]]) {
        return (NSSegmentedControl *)object;
    }
    return nil;
}

ctd_status ctd_items_clear(ctd_handle widget) {
    id object = ctd_resolve(widget);
    if (!object) return CTD_ERR_STALE;
    // A list is not a property and does not come through ctd_set_*, and a
    // segmented control with three segments is wider than one with none.
    ctd_forget_size(widget);
    NSSegmentedControl *bar = ctd_item_bar(object);
    if (bar) { [bar setSegmentCount:0]; return CTD_OK; }
    NSPopUpButton *menu = ctd_item_list(object);
    if (!menu) return CTD_ERR_KIND;
    [menu removeAllItems];
    return CTD_OK;
}

ctd_status ctd_items_add(ctd_handle widget, const char *utf8, int32_t len) {
    if (ctd_has_nul(utf8, len)) return CTD_ERR_RANGE;
    id object = ctd_resolve(widget);
    if (!object) return CTD_ERR_STALE;
    ctd_forget_size(widget);
    NSSegmentedControl *bar = ctd_item_bar(object);
    if (len < 0) return CTD_ERR_RANGE;
    if (bar) {
        NSInteger at = [bar segmentCount];
        [bar setSegmentCount:at + 1];
        [bar setLabel:ctd_string(utf8, len) forSegment:at];
        return CTD_OK;
    }
    NSPopUpButton *menu = ctd_item_list(object);
    if (!menu) return CTD_ERR_KIND;
    NSString *text = ctd_string(utf8, len);
    // NSPopUpButton drops a duplicate title, which would silently renumber
    // every later index and make a selection point at the wrong row. Adding
    // the item directly keeps the list exactly as the caller wrote it.
    NSMenuItem *item = [[NSMenuItem alloc] initWithTitle:text
                                                 action:NULL
                                          keyEquivalent:@""];
    [[menu menu] addItem:item];
    [item release];
    return CTD_OK;
}

ctd_status ctd_items_count(ctd_handle widget, int32_t *out) {
    id object = ctd_resolve(widget);
    if (!object) return CTD_ERR_STALE;
    NSSegmentedControl *bar = ctd_item_bar(object);
    if (bar) {
        if (out) *out = (int32_t)[bar segmentCount];
        return CTD_OK;
    }
    NSPopUpButton *menu = ctd_item_list(object);
    if (!menu) return CTD_ERR_KIND;
    if (out) *out = (int32_t)[menu numberOfItems];
    return CTD_OK;
}

int32_t ctd_items_at(ctd_handle widget, int32_t index, char *out, int32_t cap) {
    id object = ctd_resolve(widget);
    if (!object) return CTD_ERR_STALE;
    NSSegmentedControl *bar = ctd_item_bar(object);
    if (bar) {
        if (index < 0 || index >= (int32_t)[bar segmentCount]) return CTD_ERR_RANGE;
        NSString *label = [bar labelForSegment:index];
        return ctd_copy_out(label ? label : @"", out, cap);
    }
    NSPopUpButton *menu = ctd_item_list(object);
    if (!menu) return CTD_ERR_KIND;
    if (index < 0 || index >= (int32_t)[menu numberOfItems]) return CTD_ERR_RANGE;
    return ctd_copy_out([[menu itemAtIndex:index] title], out, cap);
}
